File: ssl/src/io.c
/*

Copyright (C) 1998-2001 Eelco Visser <visser@acm.org>

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2, or (at your option)
any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
02111-1307, USA.

*/

#include <stdlib.h>
#include <stratego.h>

/* IO */

ATermTable SSL_file_table = NULL;

void SSL_file_table_init(void)
{
  SSL_file_table = ATtableCreate(16, 80);
  ATtablePut(SSL_file_table, ATmake("stdin"),  ATmake("<int>", (int)stdin));
  ATtablePut(SSL_file_table, ATmake("stdout"), ATmake("<int>", (int)stdout));
  ATtablePut(SSL_file_table, ATmake("stderr"), ATmake("<int>", (int)stderr));
}

ATerm SSL_file_exists(ATerm t)
{
  char *name;
  FILE *file;
  name = t_string(t);

  if((file = fopen(name, "r")) == NULL)
    _fail(t);
  else
    { 
      fclose(file);
      return(t);
    }
  return(t);
}

ATerm SSL_open_file(ATerm name, ATerm mode)
{
  FILE *file;

  if(SSL_file_table == NULL)
    SSL_file_table_init();

  if(ATmatch(name, "stdout"))
    file = stdout;
  else if(ATmatch(name, "stderr"))
    file = stderr;
  else if(ATmatch(name, "stdin"))
    file = stdin;
  else if(t_is_string(name) && t_is_string(mode))
    {
      file = fopen(t_string(name), t_string(mode));
      if(file == NULL)
	_fail(name);
      ATtablePut(SSL_file_table, name, ATmake("<int>", (int)file));      
    }
  return(name);
}

// obsolete: use SSL_open_file(name, "a")

ATerm SSL_append_file(ATerm name)
{
  FILE *file;

  if(SSL_file_table == NULL)
    SSL_file_table_init();
  
  if(ATmatch(name, "stdout"))
    file = stdout;
  else if(ATmatch(name, "stderr"))
    file = stderr;
  else if(ATmatch(name, "stdin"))
    file = stdin;
  else if(t_is_string(name))
    {
      file = fopen(t_string(name), "a");
      if(file == NULL)
	_fail(name);
      ATtablePut(SSL_file_table, name, ATmake("<int>", (int)file));      
    }
  return(name);
}

FILE *_SSL_file_table_lookup(ATerm name)
{
  ATerm desc;

  /* ATfprintf(stderr, "_SSL_file_table_lookup(%t)\n", name); */

  if(SSL_file_table == NULL)
    SSL_file_table_init();
  
  if((desc = ATtableGet(SSL_file_table, name)) == NULL)
    return NULL;
  else
    return (FILE *)ATgetInt((ATermInt)desc);
}

ATerm SSL_close_file(ATerm name)
{
  FILE *file;

  if(SSL_file_table == NULL)
    SSL_file_table_init();
  
  if(ATmatch(name, "stdout"))
    fclose(stdout);
  else if(ATmatch(name, "stderr"))
    fclose(stderr);
  else if(ATmatch(name, "stdin"))
    fclose(stdin);
  else if((file = _SSL_file_table_lookup(name)) != NULL)
    fclose(file);
  else
    _fail(name);

  return(name);
}

ATerm SSL_print(ATerm file, ATerm str)
{  
  FILE *outfile;

  outfile = _SSL_file_table_lookup(file);
  if(outfile == NULL) 
    _fail(file);
  str = consnil_to_list(str);
  while(!ATisEmpty((ATermList)str))
    {
      if(ATisString(ATgetFirst((ATermList)str)))
	ATfprintf(outfile, "%s", t_string(ATgetFirst((ATermList)str)));
      else if(ATisReal(ATgetFirst((ATermList)str)))
	ATfprintf(outfile, "%.2f", ATgetReal((ATermReal)ATgetFirst((ATermList)str)));
      else
	ATfprintf(outfile, "%t", ATgetFirst((ATermList)str));
      str = (ATerm)ATgetNext((ATermList)str);
    }
  return(str);
}

ATerm SSL_printnl(ATerm file, ATerm str)
{  
  FILE *outfile;

  outfile = _SSL_file_table_lookup(file);
  if(outfile == NULL) {
    ATfprintf(stderr, "printnl: could not open file: %t\n", file);
    _fail(file);
  }
  str = consnil_to_list(str);
  if(!(ATgetType(str) == AT_LIST)) {
    ATfprintf(stderr, "ST_printnl: argument not a list: %t\n", str);
    _fail(str);
  }
  while(!ATisEmpty((ATermList)str))
    {
      if(ATisString(ATgetFirst((ATermList)str)))
	ATfprintf(outfile, "%s", t_string(ATgetFirst((ATermList)str)));
      else if(ATisReal(ATgetFirst((ATermList)str)))
	ATfprintf(outfile, "%.2f", ATgetReal((ATermReal)ATgetFirst((ATermList)str)));
      else
	ATfprintf(outfile, "%t", ATgetFirst((ATermList)str));
      str = (ATerm)ATgetNext((ATermList)str);
    }
  ATfprintf(outfile, "\n");
  return(str);
}

ATerm SSL_printascii(ATerm file, ATerm str)
{   
  FILE *outfile;

  /* ATfprintf(stderr, "SSL_printascii(): top = %t\n", Ttop()); */

  outfile = _SSL_file_table_lookup(file);
  if(outfile == NULL) 
    _fail(file);
  str = consnil_to_list(str);
  while(!ATisEmpty((ATermList)str))
    {
      if(ATisInt(ATgetFirst((ATermList)str)))
	ATfprintf(outfile, "%c", 
		  ATgetInt((ATermInt)ATgetFirst((ATermList)str)));
      else
	ATfprintf(outfile, "%t", ATgetFirst((ATermList)str));
      str = (ATerm)ATgetNext((ATermList)str);
    }
  /* ATfprintf(outfile, "\n"); */
  return(str);
}

ATerm SSL_ReadFromFile(ATerm filename)
{
  ATerm in_term;
  FILE *infile;

  if(ATmatch(filename, "stdin"))
    infile = stdin;
  else if(ATisString(filename))
    infile = fopen(t_string(filename), "r");
  else
    _fail(filename);

  if(infile == NULL)
    _fail(filename);

  if((in_term = ATreadFromFile(infile)) == NULL)
    {
      ATfprintf(stderr, "not a valid term\n");
      _fail(filename);
    }

  fclose(infile);

  /* ATfprintf(stderr, "in_term = %t\n", in_term); */
  
  return(list_to_consnil(in_term));
}

ATerm SSL_WriteToFile(ATbool binary, ATerm filename, ATerm trm)
{  
  FILE *outfile;

  if(ATmatch(filename, "stdout"))
    outfile = stdout;
  else if(ATmatch(filename, "stderr"))
    outfile = stderr;
  else if(ATisString(filename))
    outfile = fopen(t_string(filename), "w");
  else 
    _fail(filename);

  if(outfile != NULL)
    {
      if(binary)
	ATwriteToBinaryFile(consnil_to_list(trm), outfile);
      else 
	{
	  ATwriteToTextFile(consnil_to_list(trm), outfile);
	  fprintf(outfile, "\n");
	}
      fclose(outfile);
    }
  return trm;
}

ATerm SSL_WriteToBinaryFile(ATerm file, ATerm t)
{
  return SSL_WriteToFile(ATtrue, file, t);
}


ATerm SSL_WriteToTextFile(ATerm file, ATerm t)
{
  return SSL_WriteToFile(ATfalse, file, t);
}

ATerm SSL_getchar(ATerm filename)
{
  ATerm in_term;
  FILE *infile;

  if((infile = _SSL_file_table_lookup(filename)) == NULL)
    {
      ATfprintf(stderr, "file %t not open\n", filename);
      _fail(filename);
    }
  else
    {
      int c;
      c = fgetc(infile);
      if(c == EOF)
	_fail(filename);
      else
	return ATmakeInt(c);
    }
  return NULL;
}

File: ssl/spec/io.r
\literate[io]

	\begin{abstract}

	A transformation system needs to input terms to transform
	and output transformed terms. Instead of providing a single
	fixed mechanism for IO, this module
	defines primitives for file input and output of
	terms and strings. These primitives can be used in a variety
	of ways to define customized IO.

	\end{abstract}

% Copyright (C) 1998-2001 Eelco Visser <visser@acm.org>
% 
% This program is free software; you can redistribute it and/or modify
% it under the terms of the GNU General Public License as published by
% the Free Software Foundation; either version 2, or (at your option)
% any later version.
% 
% This program is distributed in the hope that it will be useful,
% but WITHOUT ANY WARRANTY; without even the implied warranty of
% MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
% GNU General Public License for more details.
% 
% You should have received a copy of the GNU General Public License
% along with this program; if not, write to the Free Software
% Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
% 02111-1307, USA.

	A compiled Stratego specification applies the strategy
	\verb|main| to the command line options that it gets. When
	interpreting these it will probably be necessary to read in
	a term from file and later write the transformed term back
	to (another) file. This module provides the primitives for
	doing file input and output. Module \verb|options| defines
	strategies to parse and analyze the command line options.

\begin{code}
module io
imports list string tuple integers time exec
\end{code}

	A \verb|file| can be a string or one of the terms
	\verb|stdin|, \verb|stdout|, \verb|stderr|.

\begin{code}
signature
  sorts File
  constructors
    stdin  : File
    stdout : File
    stderr : File
\end{code}

	\verb|<ReadFromFile> file| reads the term in \verb|file|. The
	\verb|file| needs to be in textual or binary ATerm format.

	\verb|<WriteToTextFile> (file, term)| writes \verb|term| to
	file in textual ATerm format.

	\verb|<WriteToBinaryFile> (file, term)| writes \verb|term| to
	file in BAF format.

 	\verb|<print> (file, [t1,...,tn])| prints the terms \verb|ti|
	to file. If \verb|ti| is a string it is printed without quotes,
	otherwise it is printed as a term. \verb|printnl| has the same
	behaviour, but also prints a newline after \verb|tn|.

	Before printing to a file the file should be opened using
	\verb|<open-file> filename|, which truncates the file, or creates
	it if it doesn't exist.  To append to a file, open the file
	with \verb|<append-file> filename|.  The file is created if it
	doesn't exist.

\begin{code}
strategies

  print             = ?(name, strs); where(prim("SSL_print", name, strs))
  printnl           = ?(name, strs); where(prim("SSL_printnl", name, strs))
  printascii        = ?(name, strs); where(prim("SSL_printascii", name, strs))

  file-exists	    = ?name; prim("SSL_file_exists", name)
  open-file	    = ?name; prim("SSL_open_file", name)
  append-file	    = ?name; prim("SSL_append_file", name)
  close-file	    = ?name; prim("SSL_close_file", name)

  ReadFromFile      = ?file; prim("SSL_ReadFromFile", file)
  WriteToBinaryFile = ?(file, t); prim("SSL_WriteToBinaryFile", file, t)
  WriteToTextFile   = ?(file, t); prim("SSL_WriteToTextFile", file, t)
\end{code}

 	The primitive \verb|print-stack| prints the top n elements of
 	the stack if applied as \verb|<print-stack>| n or the entire
	stack if
 	applied to a non-integer term.

\begin{code}
  print-stack = 
    obsolete(!"print-stack; use debug or gdb")
\end{code}

	The strategy \verb|debug| prints the
	current term to \verb|stderr| without changing it. This is
	a useful strategy for debugging specifications (hence its
	name).

\begin{code}
strategies

  open(file) = file; ReadFromFile

  save(file) = split(file, id); WriteToTextFile

  debug      = where(split(!stderr, \x -> [x]\); printnl)
  debug(msg) = where(split(!stderr, \x -> [<msg>(),x]\); printnl)

  say(msg)   = where(msg; debug)

  echo       = where(split(!stdout, \x -> [x]\); printnl)
  echo(msg)  = where(split(!stdout, \x -> [<msg>(),x]\); printnl)

  debug-stdout(msg) = 
    obsolete(!"debug-stdout; use echo");
    where(split(!stdout, \x -> [<msg>(),x]\); printnl)

  trace(msg,s) =
    debug(msg); (s; debug(!"succeeded: ") <+ debug(!"failed: "))

  error = where(split(!stderr, id); printnl)

  fatal-error = where(error; <exit> 1)

  giving-up   = <fatal-error>["giving-up"]

  printchar   = where(split(!stdout, \x -> [x]\ ); printascii)

  printstring = where(split(!stdout, \x -> [x]\ ); print)

  print-strings-nl(out) = where(split(out, id); printnl)

  obsolete(msg) = where(msg; debug(!"obsolete library strategy: "))
\end{code}

	The operator \verb|stdio| implements a simple user-interface
	for transformers. A term is read from standard input,
	transformed with parameter strategy \verb|s| and then written
	to standard output. If the transformation failed the text
	\verb|rewriting failed| is written to standard error.

\begin{code}
strategies

  stdio(s) = <ReadFromFile> stdin;
             s; 
             split(!stdout, id); WriteToTextFile
          <+ <fatal-error> ["** rewriting failed"]
\end{code}

	A variant of this strategy provides a pair of the command-line
	options and the input file to the strategy.

\begin{code}
strategies

  stdioO(s) = split(id, <ReadFromFile> stdin); 
              s;
              split(!stdout, id); WriteToTextFile
           <+ <fatal-error> ["** rewriting failed"]
\end{code}

	Generate a new, not existing file name.

\begin{code}
strategies

  new-file =
    rec x(<conc-strings> (<new>(), ".tmp"); try(file-exists; x))

  temp-file(s) =
  where(new-file => f)
  ; split(!f, id)
  ; (s; where(<rm-files> [f])
     <+ where(<rm-files> [f]); fail)
\end{code}
